projects/newprofile_ping_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,518 @@
+
+
+```python
+deduped_clientid = dedupe(deduped_docid, "clientId")
+total_duplicates_clientid = deduped_docid_count - deduped_clientid.count()
+print("Duplicate pings percentage (by client id): {:.2f}%".format(pct(total_duplicates_clientid, deduped_docid_count)))
+```
+    Duplicate pings percentage (by client id): 0.89%
+
+
+That's disappointing: it looks like we're receiving multiple `new-profile` pings for some clients. Let's dig into this by analysing the set of pings deduped by document id. To have a clearer picture of the problem, let's make sure to aggregate the duplicates ordered by the time they were created on the client.
+
+
+```python
+# Builds an RDD with (<client id>, [<ordered reason>, <ordered reason>, ...])
+clients_with_dupes = deduped_docid.map(lambda p: (p["clientId"], [(p["payload/reason"], p["meta/creationTimestamp"])]))\
+                                  .reduceByKey(lambda a, b: sorted(a + b, key=lambda k: k[1]))\
+                                  .filter(lambda p: len(p[1]) > 1)\
+                                  .map(lambda p: (p[0], [r[0] for r in p[1]]))
+
+# Check how often each case occurs.
+sorted(clients_with_dupes.map(lambda p: tuple(p[1])).countByValue().items(), key=lambda k: k[1], reverse=True)
+```
+
+
+
+
+    [((u'shutdown', u'shutdown'), 7),
+     ((u'shutdown', u'startup'), 4),
+     ((u'startup', u'startup'), 3),
+     ((u'startup', u'shutdown'), 3),
+     ((u'shutdown', u'shutdown', u'shutdown'), 2),
+     ((u'shutdown', u'shutdown', u'startup'), 1),
+     ((u'shutdown', u'startup', u'shutdown'), 1),
+     ((u'shutdown',
+       u'shutdown',
+       u'shutdown',
+       u'shutdown',
+       u'shutdown',
+       u'shutdown'),
+      1),
+     ((u'shutdown', u'shutdown', u'shutdown', u'startup', u'shutdown'), 1)]
+
+
+
+
+```python
+collected_clientid_reasons = clients_with_dupes.collect()
+```
+
+```python
+num_shutdown_dupes = sum([len(filter(lambda e: e == 'shutdown', t[1])) for t in collected_clientid_reasons])
+print("Duplicate 'shutdown' pings percentage (by client id): {:.2f}%"\
+      .format(pct(num_shutdown_dupes, ping_count)))
+```
+    Duplicate 'shutdown' pings percentage (by client id): 1.07%
+
+
+The multiple pings we're receiving for the same client id are mostly `new-profile` pings with reason `shutdown`. This is not too surprising, as most of the `new-profile` pings are being sent at shutdown (75%).
+
+But does the pingsender have anything to do with this? Let's attack the problem like this:
+
+- get a list of the "misbehaving" clients;
+- take a peek at their pings (redact client ids/ids);
+- figure out the next steps.
+
+
+```python
+sorted_clientid_dupes = sorted(collected_clientid_reasons, key=lambda k: len(k[1]), reverse=True)
+misbehaving_clients = [client_id for client_id, _ in sorted_clientid_dupes]
+to_redact = subset.filter(lambda p: p.get("clientId") == misbehaving_clients[1]).collect()
```

Comment:
So you're only looking at pings that are from the 2nd misbehaving client? Shouldn't it be `p.get("clientId") in misbehaving_clients`?